sway-core/src/semantic_analysis/const_fold.rs

```diff
@@ -0,0 +1,264 @@
+use super::{
+    TypedAstNode, TypedAstNodeContent, TypedExpression, TypedExpressionVariant,
+    TypedStructExpressionField,
+};
+
+use crate::semantic_analysis::declaration::ProjectionKind;
+use std::collections::HashMap;
+use sway_ir::{
+    constant::{Constant, ConstantValue},
+    context::Context,
+    module::Module,
+};
+use sway_types::ident::Ident;
+
+// A HashMap that can hold multiple values and
+// fetch values in a LIFO manner. Rust's MultiMap
+// handles values in a FIFO manner.
+pub struct MappedStack<K: std::cmp::Eq + std::hash::Hash, V> {
+    container: HashMap<K, Vec<V>>,
+}
+
+impl<K: std::cmp::Eq + std::hash::Hash, V> MappedStack<K, V> {
+    pub fn new() -> MappedStack<K, V> {
+        MappedStack {
+            container: HashMap::<K, Vec<V>>::new(),
+        }
+    }
+    pub fn push(&mut self, k: K, v: V) {
+        match self.container.get_mut(&k) {
+            Some(val_vec) => {
+                val_vec.push(v);
+            }
+            None => {
+                self.container.insert(k, vec![v]);
+            }
+        }
+    }
+    pub fn get(&self, k: &K) -> Option<&V> {
+        self.container.get(k).and_then(|val_vec| val_vec.last())
+    }
+    pub fn pop(&mut self, k: &K) {
+        match self.container.get_mut(k) {
+            Some(val_vec) => {
+                val_vec.pop();
+                if val_vec.is_empty() {
+                    self.container.remove(k);
+                }
```

Comment:
It wouldn't hurt to just leave the empty vec?

Reply:
> It wouldn't hurt to just leave the empty vec?

That depends. I was thinking that this data structure could someday be useful elsewhere too (where its size would matter), and in which case it's better for us to remove non-existing keys (and reclaim unused memory). I don't have a strong opinion on this, so if you think that removing this code (i.e, not deleting the key) is better, I'll do that.